src/Indigo/Contracts/HomebaseLite/Types/Self.hs

```diff
@@ -0,0 +1,84 @@
+-- SPDX-FileCopyrightText: 2022 Tezos Commons
+-- SPDX-License-Identifier: LicenseRef-MIT-TC
+
+module Indigo.Contracts.HomebaseLite.Types.Self
+  ( Parameter(..)
+  , Storage(..)
+  , Configuration(..)
+  , URI(..)
+  , Seconds(..)
+  , ProposalInfo(..)
+  , FA2Config(..)
+  ) where
+
+import Indigo
+
+import Lorentz.Contracts.Spec.FA2Interface (BalanceResponseItem, TokenId)
+
+data Parameter
+  = Set_admin Address
+  | Accept_admin ()
+  | Add_maintainers [Address]
+  | Remove_maintainers [Address]
+  | Configure Configuration
+  | Propose ("proposal_uri" :! URI, "choices" :! [MText])
+  | Vote ("proposal_uri" :! URI, "choice_index" :! Natural)
+  | Verify_min_balance [BalanceResponseItem]
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue)
+
+[entrypointDoc| Parameter plain |]
+
+newtype Seconds = Seconds {unSeconds :: Natural}
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+[typeDoc| Seconds "A natural representing a number of seconds."|]
+
+data Configuration = Configuration
+  { cExpireTime :: Seconds
+  , cVoteDelay :: Seconds
+  , cQuorumThreshold :: Natural
+  , cMinimumBalance :: Natural
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+[typeDoc| Configuration "Options defining the behaviour and life-cycle of proposals."|]
+
+data ProposalInfo = ProposalInfo
+  { piLevel :: Natural
+  , piStartsAt :: Timestamp
+  , piExpiresAt :: Timestamp
+  , piQuorumThreshold :: Natural
+  , piChoices :: [MText]
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+[typeDoc| ProposalInfo "Information defining a proposal."|]
+
+newtype URI = URI MText
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+[typeDoc| URI "Text representing IPFS URI for a proposal."|]
+
+data Storage = Storage
+  { sAdmin :: Address
+  , sAdminCandidate :: Address
+  , sMaintainers :: BigMap Address ()
+  , sConfiguration :: Configuration
+  , sProposals :: BigMap ("proposal_uri" :! URI) ProposalInfo
+  , sVotes :: BigMap ("proposal_uri" :! URI, "voter_address" :! Address)
+      ("vote_choice" :! Natural)
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+[typeDoc| Storage "Contract storage."|]
+
+data FA2Config = FA2Config
+  { fa2Addr :: Address
+  , fa2TokenId :: TokenId
+  }
```

Comment:
Considering that, as per spec:
> The address and token_id of the governance token must be provided at origination and can never be changed.

this `FA2Config` should be part of the `Storage` IMO.

The way that you've integrated into it has the disadvantages that:
- it requires (re-)compilation from Indigo to Michelson before generation. IOW it's a compile-time option.
This might sound not so bad, but most apps will want to be able to use the Michelson contract directly (including `homebase`, or at least this was the case for `baseDAO`), and not depend on `Indigo` (or `Haskell`) to be able to originate the contract.
- it's harder to retrieve this info from an indexer, given that it will be buried in the `code` rather then organized in a `storage` field.

The only advantages that I can tell is a slight cost reduction, which is fair but probably not that substantial.

Regardless, as I said I'm afraid that for `homebase` this is a requirement and that indeed these values need to be provided at _origination_ and not _compilation_

Reply:
Ah, ok. This makes a few things easier, actually. Done. Evidently, I've misunderstood the spec.

Reply:
no problem, TBF it could have been made more clear